nhrpd/nhrp_interface.c

```diff
@@ -19,9 +19,47 @@
 
 #include "nhrpd.h"
 #include "os.h"
-#include "netlink.h"
+#include "hash.h"
 
 DEFINE_MTYPE_STATIC(NHRPD, NHRP_IF, "NHRP interface");
+DEFINE_MTYPE_STATIC(NHRPD, NHRP_IF_GRE, "NHRP GRE interface");
+
+struct hash *nhrp_gre_list;
+
+static unsigned int nhrp_gre_info_key(const void *data)
+{
+	const struct nhrp_gre_info *r = data;
+
+	return r->ifindex;
+}
+
+static bool nhrp_gre_info_cmp(const void *data, const void *key)
+{
+	const struct nhrp_gre_info *a = data, *b = key;
+
+	if (a->ifindex == b->ifindex)
+		return true;
+	return false;
+}
+
+static void *nhrp_interface_gre_alloc(void *data)
+{
+	struct nhrp_gre_info *a;
+	struct nhrp_gre_info *b = data;
+
+	a = XMALLOC(MTYPE_NHRP_IF_GRE, sizeof(struct nhrp_gre_info));
+	memcpy(a, b, sizeof(struct nhrp_gre_info));
+	return a;
+}
+
+struct nhrp_gre_info *nhrp_gre_info_alloc(struct nhrp_gre_info *p)
+{
+	struct nhrp_gre_info *a;
+
+	a = (struct nhrp_gre_info *)hash_get(nhrp_gre_list, p,
+					     nhrp_interface_gre_alloc);
+	return a;
+}
 
 static void nhrp_interface_update_cache_config(struct interface *ifp,
```

Comment:
prototype in a weird place - move it to the beginning of the file?

Reply:
ok. added a commit log